runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/ProjectionPushdownOptimizerTest.java

```diff
@@ -0,0 +1,364 @@
+        .apply(new FieldAccessTransform(FieldAccessDescriptor.withFieldNames("qux")));
+
+    MultipleOutputSourceWithPushdown expectedSource =
+        new MultipleOutputSourceWithPushdown(
+            FieldAccessDescriptor.withFieldNames("foo", "bar"),
+            FieldAccessDescriptor.withFieldNames("baz", "qux"));
+
+    ProjectionPushdownOptimizer.optimize(p);
+    Assert.assertTrue(pipelineHasTransform(p, expectedSource));
+    Assert.assertFalse(pipelineHasTransform(p, originalSource));
+  }
+
+  private static boolean pipelineHasTransform(Pipeline p, PTransform<?, ?> t) {
+    HasTransformVisitor hasTransformVisitor = new HasTransformVisitor(t);
+    p.traverseTopologically(hasTransformVisitor);
+    return hasTransformVisitor.found;
+  }
+
+  private static class HasTransformVisitor extends Defaults {
+    private final PTransform<?, ?> t;
+    boolean found = false;
+
+    HasTransformVisitor(PTransform<?, ?> t) {
+      this.t = t;
+    }
+
+    @Override
+    public CompositeBehavior enterCompositeTransform(Node node) {
+      if (t.equals(node.getTransform())) {
+        found = true;
+        return CompositeBehavior.DO_NOT_ENTER_TRANSFORM;
+      }
+      return CompositeBehavior.ENTER_TRANSFORM;
+    }
+  }
+
+  private static class FieldAccessTransform extends PTransform<PCollection<Row>, PCollection<Row>> {
+    private final FieldAccessDescriptor fieldAccessDescriptor;
+
+    FieldAccessTransform(FieldAccessDescriptor fieldAccessDescriptor) {
+      this.fieldAccessDescriptor = fieldAccessDescriptor;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollection<Row> input) {
+      return input
+          .apply(
+              ParDo.of(
+                  new DoFn<Row, Row>() {
+                    @SuppressWarnings("unused") // used by reflection
+                    @FieldAccess("row")
+                    private final FieldAccessDescriptor fieldAccessDescriptor =
+                        FieldAccessTransform.this.fieldAccessDescriptor;
+
+                    @ProcessElement
+                    public void processElement(
+                        @FieldAccess("row") Row row, OutputReceiver<Row> outputReceiver)
+                        throws Exception {
+                      // Do nothing; we don't need to execute this DoFn.
+                    }
+                  }))
+          .setRowSchema(createStringSchema(fieldAccessDescriptor));
+    }
+  }
+
+  private static Schema createStringSchema(FieldAccessDescriptor fieldAccessDescriptor) {
+    if (fieldAccessDescriptor.getAllFields()) {
+      return createStringSchema(FieldAccessDescriptor.withFieldNames("foo", "bar", "baz", "qux"));
+    }
+    Schema.Builder schemaBuilder = Schema.builder();
+    for (FieldDescriptor field : fieldAccessDescriptor.getFieldsAccessed()) {
+      schemaBuilder.addStringField(field.getFieldName());
+    }
+    return schemaBuilder.build();
+  }
+
+  private abstract static class SchemaTransform<InputT extends PInput>
```

Comment:
nit: maybe clarify this is for transforms that produce a schema'd PCollection
```suggestion
  private abstract static class SchemaSourceTransform<InputT extends PInput>
```

Reply:
Done.